cardano-cli/src/Cardano/CLI/Shelley/Run/Key.hs

```diff
@@ -415,86 +421,198 @@ runConvertITNBip32ToStakeKey (ASigningKeyFile (SigningKeyFile sk)) (OutputFile o
-decodeBech32Key :: Text
-                -> Either ConversionError
-                          (Bech32.HumanReadablePart, Bech32.DataPart, ByteString)
-decodeBech32Key key =
-  case Bech32.decodeLenient key of
-    Left err -> Left $ Bech32DecodingError key err
-    Right (hRpart, dataPart) -> case Bech32.dataPartToBytes dataPart of
-                                  Nothing -> Left $ ITNError hRpart dataPart
-                                  Just bs -> Right (hRpart, dataPart, bs)
-
-readFileITNKey :: FilePath -> IO (Either ConversionError Text)
+    Nothing -> Left $ ItnSigningKeyDeserialisationError privkeyBS
+
+readFileITNKey :: FilePath -> IO (Either ItnKeyConversionError Text)
 readFileITNKey fp = do
   eStr <- Exception.try $ readFile fp
   case eStr of
-    Left e -> return . Left $ Bech32ReadError fp e
+    Left e -> return . Left $ ItnReadBech32FileError fp e
     Right str -> return . Right . Text.concat $ Text.words str
+
+--------------------------------------------------------------------------------
+-- `cardano-address` extended signing key conversions
+--------------------------------------------------------------------------------
+
+runConvertCardanoAddressSigningKey
+  :: CardanoAddressKeyType
+  -> SigningKeyFile
+  -> OutputFile
+  -> ExceptT ShelleyKeyCmdError IO ()
+runConvertCardanoAddressSigningKey keyType skFile (OutputFile outFile) = do
+  sKey <- firstExceptT ShelleyKeyCmdCardanoAddressSigningKeyFileError
+    . newExceptT
+    $ readCardanoAddressSigningKeyFile keyType skFile
+  firstExceptT ShelleyKeyCmdWriteFileError . newExceptT
+    $ writeSomeCardanoAddressSigningKeyFile outFile sKey
+
+-- | Some kind of signing key that was converted from a @cardano-address@
+-- signing key.
+data SomeCardanoAddressSigningKey
+  = ACardanoAddrShelleyPaymentSigningKey !(SigningKey PaymentExtendedKey)
+  | ACardanoAddrShelleyStakeSigningKey !(SigningKey StakeExtendedKey)
+  | ACardanoAddrByronSigningKey !(SigningKey ByronKey)
+
+-- | An error that can occur while converting a @cardano-address@ extended
+-- signing key.
+data CardanoAddressSigningKeyConversionError
+  = CardanoAddressSigningKeyBech32DecodeError !Bech32DecodeError
+  -- ^ There was an error in decoding the string as Bech32.
+  | CardanoAddressSigningKeyDeserialisationError !ByteString
+  -- ^ There was an error in converting the @cardano-address@ extended signing
+  -- key.
+  deriving (Show, Eq)
+
+instance Error CardanoAddressSigningKeyConversionError where
+  displayError = Text.unpack . renderCardanoAddressSigningKeyConversionError
+
+-- | Render an error message for a 'CardanoAddressSigningKeyConversionError'.
+renderCardanoAddressSigningKeyConversionError
+  :: CardanoAddressSigningKeyConversionError
+  -> Text
+renderCardanoAddressSigningKeyConversionError err =
+  case err of
+    CardanoAddressSigningKeyBech32DecodeError decErr ->
+      Text.pack (displayError decErr)
+    CardanoAddressSigningKeyDeserialisationError _bs ->
```

Comment:
Did you mean to add the `ByteString` to the error here?

Reply:
No, I purposely left it out. I think there'd be a security concern in including this private data in an error message.

Reply:
You should leave a comment stating this.

Reply:
Sure, that's fine.